Add support for the Issue Timeline API

yallop opened the pull request:
This PR adds support for the experimental [Issue Timeline API](https://developer.github.com/v3/issues/timeline/), which exposes a variety of events not available in standard issue event streams, including repository cross-references.

Experiments reveal that many of the fields, including `id`, are not returned in every case, and so the new types mark most of the fields as optional (17c9e5f).

The timeline API uses a distinct media-type (`application/vnd.github.mockingbird-preview`), and so the `get` function now accepts an additional optional argument (42fd1dc).

dave-tucker commented on lib/github_s.mli:
```diff
@@ -339,10 +339,26 @@ module type Github = sig
+          media-type used by the
+          {{:https://developer.github.com/v3/} GitHub API v3} *)
+
+      val experimental : string
```
@yallop per [the documentation](https://developer.github.com/v3/previews/) we should probably rename this as `issue_timeline` as it seems that each experimental feature has it's own media type.

dsheets replied:
I'm not sure this should be exposed in the interface as a lot of these media types are temporary.

dsheets commented on lib/github.atd:
```diff
@@ -1193,4 +1193,66 @@ type update_gist = {
   files: (string * update_gist) list <json repr="object">
 } <ocaml field_prefix="update_gist_">
 
+
```
This blank line is not necessary.

dsheets commented on lib/github.atd:
```diff
@@ -1193,4 +1193,66 @@ type update_gist = {
+  | Closed <json name="closed">
+  | Commented <json name="commented">
+  | Committed <json name="committed">
+  | Cross_referenced <json name="cross-referenced">
```
It's really neat that GitHub used dashes instead of underscores for this name and only this name.

dsheets commented on lib/github.atd:
```diff
@@ -1193,4 +1193,66 @@ type update_gist = {
+type rename = {
+  from: string;
+  to_: string <json name="to">;
+} <ocaml field_prefix="rename_">
```
Is this the same as `issue_rename`?

dsheets commented on lib/github.atd:
```diff
@@ -1193,4 +1193,66 @@ type update_gist = {
+  ?url: string option;
+  ?actor: user option;
+  ?issue: issue option;
+} <ocaml field_prefix="source_">
```
Maybe this should be given a more specific name like `timeline_source_`?

dsheets commented on lib/github.atd:
```diff
@@ -1193,4 +1193,66 @@ type update_gist = {
+type timeline_label = {
+  name: string;
+  color: string;
+} <ocaml field_prefix="timeline_label_">
```
Is this the same as `base_label`?

dsheets commented on lib/github_core.ml:
```diff
@@ -785,17 +789,23 @@ module Make(Env : Github_s.Env)(Time : Github_s.Time)(CL : Cohttp_lwt.Client)
+    struct
+      let v3 = "application/vnd.github.v3+json"
+      let experimental = "application/vnd.github.mockingbird-preview"
+    end
```
I'm not sure this is necessary as most API requests will use the normal `v3` label and the experimental labels are specific to experimental features.

dsheets commented on lib/github_core.ml:
```diff
@@ -812,7 +822,8 @@ module Make(Env : Github_s.Env)(Time : Github_s.Time)(CL : Cohttp_lwt.Client)
         (Monad.(request ?token ?params
-                  {meth; uri; headers=realize_headers headers; body })
+                  {meth; uri; headers=realize_headers headers;
+                   body })
```
This change isn't necessary. Should `media_type` be exposed here, too?

dsheets commented on lib/github_core.ml:
```diff
@@ -1509,6 +1527,14 @@ module Make(Env : Github_s.Env)(Time : Github_s.Time)(CL : Cohttp_lwt.Client)
+      let uri = URI.issue_timeline ~user ~repo ~num in
+      let headers =
+        C.Header.(add (init ()) "accept"
+                    "application/vnd.github.mockingbird-preview") in
```
Why do you set the `accept` header and use the `media_type` feature both?

dsheets commented on lib/github_s.mli:
```diff
@@ -1038,6 +1055,13 @@ module type Github = sig
+      ?token:Token.t -> user:string -> repo:string -> num:int -> unit ->
+      Github_t.timeline_event Stream.t
+
+    (** [timeline_events ~user ~repo ~num ()] is a stream of all timeline
```
Please remove the empty line between signature and doc.

yallop commented:
Thanks for the reviews!  I've addressed the comments in f38dfa5.

I haven't exposed `media_type` for methods other than `GET` for the moment, since there's no immediate need, and it can be added later without breaking things.  Other than that I've made all the suggested changes.

yallop commented:
The [Fedora build failure](https://travis-ci.org/mirage/ocaml-github/jobs/203525805) appears to be due to an unrelated problem with `dose3`.

avsm commented:
Fedora pod2html error should hopefully be addressed by https://github.com/ocaml/opam-repository/pull/8687

yallop commented:
Thanks, @avsm; that does seem to have fixed it.
